fairlearn/metrics/_annotated_metric_function.py

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) Fairlearn contributors.
+# Licensed under the MIT License.
+
+from typing import Callable, Dict, List, Optional
+
+import logging
+import numpy as np
+import pandas as pd
+
+logger = logging.getLogger(__name__)
+
+_DEFAULT_NAME = "metric"
+
+_METRIC_FUNCTION_NONE = "Found 'None' instead of metric function"
+_METRIC_FUNCTION_NOT_CALLABLE = "Object passed as metric function not callable"
+
+
+class AnnotatedMetricFunction:
+    """Wraps functions to make them callable with a DataFrame argument.
+
+    The :class:`MetricFrame` makes extensive use of `pandas` DataFrames
+    internally. In particular, combinations of sensitive (and control)
+    features are selected using `DataFrame.groupby()` and acted on via
+    `DataFrame.apply()`. The net result of this is that it is useful
+    to have a function wrapper which can be invoked with a DataFrame
+    argument, and extract its arguments from that DataFrame.
+
+    Parameters
+    ----------
+    func : callable
+        The metric function we wish to invoke
+    name: str
+        An optional string defining the name of the function
+    positional_argument_names: List[str]
+        The column names to be extracted and passed as positional arguments
+        when invoking the function
+    kw_argument_mapping: Dict[str, str]
+        The column names which are to be passed as keyword arguments
+        when invoking the function. Since the DataFrame column names may
+        not match the function's argument names, this is a dictionary
+        where the keys are the function argument names, and the values
+        are the column names
+    """
+
+    def __init__(
+        self,
+        *,
+        func: Callable,
+        name: Optional[str],
+        postional_argument_names: List[str] = None,
+        kw_argument_mapping: Dict[str, str] = None
+    ):
+        if func is None:
+            raise ValueError(_METRIC_FUNCTION_NONE)
+        if not callable(func):
+            raise ValueError(_METRIC_FUNCTION_NOT_CALLABLE)
+        self.func = func
+
+        if name is None:
+            if hasattr(func, "__name__"):
+                self.name = func.__name__
+            else:
+                logger.warning("Supplied 'func' had no __name__ attribute")
```

Comment:
this is not tested (hard to find what's not tested, this is why I love codecov)

Reply:
Do you mean that I don't check for the log message? This branch of the `if` statement is covered by `test_constructor_no_name()`

Reply:
Ping @adrinjalali I'm not sure which part you mean isn't tested.

Reply:
Yes, the test doesn't check that this warning is raised (I guess that means the test actually raises a warning, which it ideally shouldn't).